File: Sources/Emo/Emo.swift
import PlatformSupport

/// A single emoji suggestion returned by ``Emo/suggestions(for:limit:skinTone:)``.
public struct EmoSuggestion: Identifiable, Sendable, Equatable {
    init(emoji: String, confidence: Double) {
        id = emoji
        self.emoji = emoji
        self.confidence = confidence
    }

    /// A stable identifier for this suggestion. This is the same value as ``emoji``.
    public let id: String

    /// The suggested emoji.
    public let emoji: String

    /// The model's normalized confidence for this suggestion, from `0...1`.
    public let confidence: Double
}

/// Errors thrown while loading or running the model. (`MessageError` is
/// `LocalizedError` wherever Foundation exists, so `localizedDescription`
/// shows `message`.)
public enum EmoError: MessageError, Sendable {
    /// A model resource (model, tokenizer, or metadata) could not be found.
    case modelNotFound
    /// On-device prediction failed or returned an unexpected output.
    case predictionFailed

    public var message: String {
        switch self {
        case .modelNotFound: "An Emo model resource was not found."
        case .predictionFailed: "On-device emoji prediction failed."
        }
    }
}

/// On-device emoji suggestion for short tasks, calendar entries, or messages.
///
/// `Emo` turns a short phrase into ranked emoji suggestions, fully on device,
/// across 23 languages. A small hashed n-gram stream plus a transformer over a
/// pruned multilingual token sequence run through the shared inference session
/// (Core ML on Apple, LiteRT elsewhere). Create one once and reuse it.
///
/// ```swift
/// let emo = Emo()
/// let suggestions = try await emo.suggestions(for: "Pay my bills")
/// // ["💰", "💳", "🧾", ...]
/// let toned = try await emo.suggestions(for: "go for a run", limit: 1, skinTone: .medium)
/// // "🏃🏽"
/// ```
public final class Emo: @unchecked Sendable {
    /// Resolve the model's assets (downloading/adopting as needed), reporting
    /// progress `0...1`.
    typealias ResolveAssets = @Sendable (@escaping @Sendable (Double) -> Void) async throws -> ModelAssets

    private let loader: LazyLoader<Model>
    private let availability: @Sendable () -> Bool

    /// Creates a suggester. Construction does no work and starts no download;
    /// the model loads on the first ``suggestions(for:limit:skinTone:)`` or
    /// ``download(progress:)``, off your calling thread.
    ///
    /// Emo is small, so with no `directory` (the default) the model **bundled
    /// into the SDK** is used - ready offline, no download. Passing a
    /// `directory` opts into adopt-or-download: if it already contains the model
    /// it is used offline; otherwise the model is downloaded into it and reused.
    /// (Disable the `BundledModel` package trait to make the default download on
    /// demand instead.)
    public convenience init(directory: String? = nil) {
        self.init(directory: directory, cacheRoot: nil)
    }

    /// Binding entry point that also supplies the platform base cache root under
    /// which the managed layout lives (the app cache dir on Android, node
    /// `~/.cache` on the web). On Apple/Linux FileManager provides it, so the
    /// public `init(directory:)` passes `nil`.
    @_spi(EmoBindings)
    public convenience init(directory: String?, cacheRoot: String?) {
        self.init(
            resolve: { try await Emo.defaultAssets(directory: directory, cacheRoot: cacheRoot, progress: $0) },
            isAvailable: { Emo.defaultIsAvailable(directory: directory, cacheRoot: cacheRoot) }
        )
    }

    /// Creates a suggester from explicitly provided assets (used by the
    /// Android/JNI and custom-deployment paths).
    @_spi(EmoBindings)
    public convenience init(assets: ModelAssets) {
        self.init(resolve: { _ in assets }, isAvailable: { true })
    }

    init(resolve: @escaping ResolveAssets, isAvailable: @escaping @Sendable () -> Bool) {
        loader = LazyLoader { progress in try Model(assets: await resolve(progress)) }
        availability = isAvailable
    }

    /// Whether the model is available for this suggester with no network:
    /// cached (for the download source), present (for a directory), or bundled.
    public func isDownloaded() -> Bool { availability() }

    /// Download and load the model ahead of time, so the first
    /// ``suggestions(for:limit:skinTone:)`` is instant. Reports download
    /// progress `0...1`. Concurrent calls, and an implicit load from a
    /// suggestion, share one download. A no-op once loaded (see
    /// ``isDownloaded()``).
    public func download(progress: @Sendable @escaping (Double) -> Void = { _ in }) async throws {
        try await loader.run(progress: progress)
    }

    /// Await model readiness. The bindings use this to surface load errors
    /// eagerly; apps can just call ``suggestions(for:limit:skinTone:)``.
    @_spi(EmoBindings)
    public func waitUntilLoaded() async throws {
        _ = try await loader.value()
    }

    /// Returns up to `limit` emoji suggestions for `text`, most likely first.
    ///
    /// - Parameters:
    ///   - text: A short task, calendar entry, note, or message draft.
    ///   - limit: The maximum number of suggestions. Pass `1` for only the best emoji.
    ///   - skinTone: Preferred skin tone for skin-tone-capable emoji.
    /// - Returns: Up to `limit` suggestions. Empty input returns `[]`.
    public func suggestions(for text: String, limit: Int = 3, skinTone: EmojiSkinTone = .default) async throws -> [EmoSuggestion] {
        let trimmed = text.trimmed
        guard !trimmed.isEmpty else { return [] }
        let model = try await loader.value()
        return try await model.suggestions(for: trimmed, limit: limit, skinTone: skinTone)
    }
}

private extension String {
    /// Trim ASCII/Unicode whitespace without Foundation (absent on Android).
    var trimmed: String {
        let scalars = unicodeScalars
        var start = scalars.startIndex
        var end = scalars.endIndex
        while start < end, scalars[start].properties.isWhitespace { start = scalars.index(after: start) }
        while end > start {
            let prev = scalars.index(before: end)
            if scalars[prev].properties.isWhitespace { end = prev } else { break }
        }
        return String(String.UnicodeScalarView(scalars[start..<end]))
    }
}

File: packages/emo-kotlin/src/main/kotlin/ai/desertant/emo/Emo.kt
package ai.desertant.emo

import ai.desertant.core.FfiReader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/** A single emoji suggestion returned by [Emo.suggestions]. */
data class EmoSuggestion(
    /** The suggested emoji. */
    val emoji: String,
    /** The model's normalized confidence for this suggestion, from `0.0` to `1.0`. */
    val confidence: Double,
)

/** Thrown when the model cannot be created, loaded, or run. */
class EmoException(message: String) : Exception(message)

/**
 * On-device multilingual emoji suggestion. Mirrors the iOS/Swift SDK: create one
 * `Emo` and reuse it; the model loads lazily on the first [suggestions] (or
 * eagerly via [download]).
 *
 * ```kotlin
 * val emo = Emo(context)                          // download on demand, cached
 * val suggestions = emo.suggestions("Pay my bills")
 * val toned = emo.suggestions("go for a run", limit = 1, skinTone = EmojiSkinTone.MEDIUM)
 * emo.close()
 * ```
 */
class Emo private constructor(private val handle: Long) : AutoCloseable {
    /**
     * A suggester using the bundled model by default. When [directory] is
     * supplied, that directory is treated as the model's home instead (adopt
     * files there, else download into it). Construction is cheap; the model
     * loads on the first [suggestions] (or eagerly via [download]).
     */
    constructor(context: android.content.Context, directory: String? = null)
        : this(if (directory == null) bundledHandleOrNull() ?: createHandle(context.cacheDir.absolutePath, null)
               else createHandle(context.cacheDir.absolutePath, directory))

    companion object {
        /**
         * A suggester using the bundled model (no network). The main Emo AAR
         * depends on the resources artifact by default because the model is
         * small; this remains useful for explicit offline construction.
         */
        fun bundled(): Emo {
            val handle = bundledHandleOrNull() ?: throw EmoException(
                "bundled model unavailable; make sure `ai.desertant:emo-tflite-resources` is present")
            return Emo(handle)
        }

        private fun bundledHandleOrNull(): Long? {
            EmoNative.ensureLoaded()
            val meta = resourceOrNull("emo_meta.json") ?: return null
            val tokenizer = resourceOrNull("emo_tokenizer.bin") ?: return null
            val model = resourceOrNull("emo.tflite") ?: return null
            val handle = EmoNative.createBundled(meta, tokenizer, model)
            return handle.takeIf { it != 0L }
        }

        private fun createHandle(cacheRoot: String, directory: String?): Long {
            EmoNative.ensureLoaded()
            val handle = EmoNative.create(
                cacheRoot.toByteArray(Charsets.UTF_8), directory?.toByteArray(Charsets.UTF_8))
            if (handle == 0L) throw EmoException("failed to create Emo")
            return handle
        }

        private fun resourceOrNull(name: String): ByteArray? =
            Emo::class.java.getResourceAsStream("/$name")?.use { it.readBytes() }
    }

    /** Whether the model is available for this suggester with no network. */
    fun isDownloaded(): Boolean = EmoNative.isDownloaded(handle) != 0

    /**
     * Download the model ahead of time so the first [suggestions] is instant. A
     * no-op once available (see [isDownloaded]). Suspends on a background
     * dispatcher.
     */
    suspend fun download(): Unit = withContext(Dispatchers.IO) {
        if (EmoNative.download(handle) != 0) throw EmoException("model download failed")
    }

    /**
     * Suggest emojis for [text], most likely first. Returns up to [limit]
     * suggestions; empty or blank input returns an empty list. Loads the model
     * lazily on first call.
     */
    suspend fun suggestions(
        text: String, limit: Int = 3, skinTone: EmojiSkinTone = EmojiSkinTone.DEFAULT,
    ): List<EmoSuggestion> = withContext(Dispatchers.Default) {
        if (text.isBlank()) return@withContext emptyList()
        val bytes = EmoNative.run(handle, text.toByteArray(Charsets.UTF_8), limit, skinTone.nativeValue)
            ?: throw EmoException("suggestion failed")
        val r = FfiReader(bytes)
        List(r.int()) { EmoSuggestion(r.string(), r.double()) }
    }

    /** Release the native model. The suggester is unusable afterwards. */
    override fun close() = EmoNative.destroy(handle)
}
